test/LibraryManager.Test/Utilities/ParallelUtilityTests.cs

```diff
@@ -0,0 +1,117 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using Microsoft.Web.LibraryManager.Utilities;
+
+namespace Microsoft.Web.LibraryManager.Test.Utilities
+{
+    [TestClass]
+    public class ParallelUtilityTests
+    {
+        [TestMethod]
+        public async Task ForEachAsync_NullAction_ShouldThrow()
+        {
+            await Assert.ThrowsExceptionAsync<ArgumentNullException>(async () =>
+                await ParallelUtility.ForEachAsync(null, 1, Array.Empty<object>())
+            );
+        }
+
+        [TestMethod]
+        public async Task ForEachAsync_0DegreesParallelism_ShouldThrow()
+        {
+            await Assert.ThrowsExceptionAsync<ArgumentOutOfRangeException>(async () =>
+                await ParallelUtility.ForEachAsync((i) => Task.CompletedTask, 0, Array.Empty<object>())
+            );
+        }
+
+        [TestMethod]
+        public async Task ForEachAsync_NegativeDegreesParallelism_ShouldThrow()
+        {
+            await Assert.ThrowsExceptionAsync<ArgumentOutOfRangeException>(async () =>
+                await ParallelUtility.ForEachAsync((i) => Task.CompletedTask, -5, Array.Empty<object>())
+            );
+        }
+
+        [TestMethod]
+        public async Task ForEachAsync_NullItems_ShouldThrow()
+        {
+            await Assert.ThrowsExceptionAsync<ArgumentNullException>(async () =>
+                await ParallelUtility.ForEachAsync((i) => Task.CompletedTask, 1, (object[])null)
+            );
+        }
+
+        [TestMethod]
+        public async Task ForEachAsync_ValidEmptyArguments_ShouldRunWithoutExceptions()
+        {
+            await ParallelUtility.ForEachAsync((i) => Task.CompletedTask, 1, Array.Empty<object>());
+        }
+
+        [TestMethod]
+        public async Task ForEachAsync_IfOneTaskFails_AllTasksGetRun()
+        {
+            int counter = 0;
+
+            Exception exception = await Assert.ThrowsExceptionAsync<Exception>(async () =>
+                await ParallelUtility.ForEachAsync((i) =>
+                {
+                    if (i % 2 == 0)
+                    {
+                        throw new Exception("Expected Failure");
+                    }
+                    counter++;
+                    return Task.CompletedTask;
+                }, 1, Enumerable.Range(1, 3))
+            );
+
+            Assert.AreEqual("Expected Failure", exception.Message);
+            Assert.AreEqual(2, counter); // both 1 and 3 ran, 2 threw the exception
+        }
+
+        [TestMethod]
+        public async Task ForEachAsync_CancellationTokenSet_DoesNotRunAllTasks()
```

Comment:
I think at least theoretically this test can fail. If you read the doc at https://docs.microsoft.com/en-us/dotnet/api/system.threading.semaphoreslim?view=netframework-4.8, it says
```
 If multiple threads are blocked, there is no guaranteed order, such as FIFO or LIFO, that controls when threads enter the semaphore.
```

That would mean that after executing the Func for "1", it could execute Func for "3" before executing it for "2".

Reply:
This test uses a `degreeOfParallelism` of 1, so "1" must completely execute before "2", which must in turn completely execute before "3".  Any different execution order would mean that the foreach iterator yielded a different input order.

If the test fails, then it would catch a change in the implementation that breaks this behavior (which seems desirable to me).

Reply:
No, "2" doesn't not have to execute completely before "3". Tasks are not necessarily awaken/given semaphore in the order they were queued up. That's what documentation is saying.  When "1" is running, "2" and "3" are blocked and queued up to acquire the semaphore. In theory (implementation-dependent), "3" could get semaphore before "2" does after "1" completes.

Reply:
It does have to execute completely.  With DoP=1, the foreach will get the first item ("1").  It will acquire the semaphore.  It launch the new task, it will reach the end of the loop and iterate (now iterating on "2").  Then it blocks on the semaphore.  It can't launch the task for "2" until the task for "1" reaches the `semaphore.Release()`.  It can't even try to start "3" until it's launched "2" (at which point, it will be waiting on the semaphore again.  Remember, the foreach itself is not parallel, it's just a serial iteration that is launching a set of tasks (and it's ultimately up to the TaskScheduler to run those in parallel).

With DoP>1 your point is entirely right.  That's why I used DoP=1 in these tests, because this way I do have absolute control of order of execution this way.

Reply:
OK, looked at that loop again, and you are right. I don't think I read it right the first time. The first item over the limit blocks the loop executing, so we never have more than one item waiting on the semaphore. For some reason I thought (incorrectly) that we queued up all items, including the ones waiting on the semaphore in the allTasks, but it's obviously not the case :) 

---
In reply to: [362246750](https://github.com/aspnet/LibraryManager/pull/538#discussion_r362246750) [](ancestors = 362246750)